src/Microsoft.Azure.SignalR/ClientInvocation/CallerClientResultsManager.cs

```diff
@@ -0,0 +1,174 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Buffers;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Diagnostics;
+using System.Collections.Generic;
+using System.Collections.Concurrent;
+using Microsoft.Azure.SignalR.Protocol;
+using Microsoft.AspNetCore.SignalR.Protocol;
+using Microsoft.AspNetCore.SignalR;
+using System.Collections;
+
+namespace Microsoft.Azure.SignalR
+{
+    internal class CallerClientResultsManager : ICallerClientResultsManager, IInvocationBinder
+    {
+        private readonly ConcurrentDictionary<string, PendingInvocation> _pendingInvocations = new();
+        private readonly ConcurrentDictionary<string, List<string>> _serviceMappingMessages = new();
+        private readonly string _serverGUID = Guid.NewGuid().ToString();
+        private long _lastInvocationId = 0;
+
+        private readonly IHubProtocolResolver _hubProtocolResolver;
+
+        public CallerClientResultsManager(IHubProtocolResolver hubProtocolResolver)
+        {
+            _hubProtocolResolver = hubProtocolResolver;
+        }
+
+        public string GenerateInvocationId(string connectionId)
+        {
+            return $"{connectionId}-{_serverGUID}-{Interlocked.Increment(ref _lastInvocationId)}";
+        }
+
+        public Task<T> AddInvocation<T>(string connectionId, string invocationId, CancellationToken cancellationToken)
+        {
+            var tcs = new TaskCompletionSourceWithCancellation<T>(
+                cancellationToken,
+                () => TryCompleteResult(connectionId, CompletionMessage.WithError(invocationId, "Canceled")));
+
+            var result = _pendingInvocations.TryAdd(invocationId, new PendingInvocation(typeof(T), connectionId, tcs, static (state, completionMessage) =>
+            {
+                var tcs = (TaskCompletionSourceWithCancellation<T>)state;
+                if (completionMessage.HasResult)
+                {
+                    tcs.SetResult((T)completionMessage.Result);
+                }
+                else
+                {
+                    tcs.SetException(new Exception(completionMessage.Error));
+                }
+            }
+            ));
+            Debug.Assert(result);
+
+            tcs.RegisterCancellation();
+
+            return tcs.Task;
+        }
+
+        public void AddServiceMappingMessage(ServiceMappingMessage serviceMappingMessage)
+        {
+            var instanceId = serviceMappingMessage.InstanceId;
+
+            lock (_serviceMappingMessages)
```

Comment:
No need to lock. `ConcurrentDictonary.AddOrUpdate()` ensures thread safe.

Reply:
fixed